### PR TITLE
ci: optimize CI workflow

### PR DESCRIPTION
**Pull request by mtrezza**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Flutter/security/policy).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/Parse-SDK-Flutter/issues?q=is%3Aissue).

### Issue Description
- move lint into separate job
- use caching for flutter download
- test only against stable flutter / dart versions (not beta, dev)

Related issue: https://github.com/parse-community/Parse-SDK-Flutter/issues/721

### Approach
Fixed issues mentioned above.

### TODOs before merging
n/a

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -11,14 +11,27 @@ on:
...
+    runs-on: ubuntu-18.04
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+      - name: Setup flutter
+        uses: subosito/flutter-action@v2
+        with:
+          cache: true
+      - name: Lint
+        run: flutter format --output=none --set-exit-if-changed packages/flutter
```
Comment (fischerscode) [inline .github/workflows/ci.yml:26]: I think we have to check the format for the dart package, too. This should only check packages/flutter.
I see 3 options:
1. Copy the job (check-lint-dart with dart-lang/setup-dart@v1)
2. use `flutter format --output=none --set-exit-if-changed packages` to format both packages
3. Install flutter and dart in one job and run both format commands

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -48,35 +48,17 @@ jobs:
...
-        run: (cd packages/dart && dart run build_runner build --delete-conflicting-outputs)
-      - name: Run tests
-        run: (cd packages/dart && dart test)
   check-flutter:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
         channel: [stable]
+    name: Test flutter (${{ matrix.os }}, ${{ matrix.channel }})
```
Comment (fischerscode) [inline .github/workflows/ci.yml:57]: Great, didn't know that's possible.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -11,52 +11,84 @@ on:
...
-      - name: Analyze project source
-        run: flutter analyze packages/flutter
       - name: Run tests
         run: (cd packages/flutter && flutter test)
+  check-dart:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+        channel: [stable]
```
Comment (fischerscode) [inline .github/workflows/ci.yml:92]: We could set strategy.fail-fast to false in order to keep the other tests running if one of the matrix fails.

Comment (mtrezza) [reply .github/workflows/ci.yml]: added

File: .github/workflows/ci.yml (hunk 2 above)
Comment (mtrezza) [reply .github/workflows/ci.yml]: Yes, makes it a bit more readable

File: .github/workflows/ci.yml (hunk 1 above)
Comment (mtrezza) [reply .github/workflows/ci.yml]: I've added a separate job, somehow feels more "correct" to let dart to its own analysis.

**Comment by mtrezza** [comment]
- I'm not happy with the lint checks, they don't output what exactly they have found. Makes it hard to fix these.
- What does the code analysis need? I thought it's only checking the source code, or does this need to run after compile?

**Comment by fischerscode** [comment]
```
Changed packages/flutter/example/test/data/repository/diet_plan/repository_diet_plan_db_test.dart
Changed packages/flutter/example/test/data/repository/diet_plan/repository_diet_plan_test.dart
Changed packages/flutter/example_livelist/lib/main.dart
Formatted 34 files (3 changed) in 0.84 seconds.
Formatting failed: 1
Error: Process completed with exit code 1.
```
They tell which files changed, I think it's more a "you missed auto formatting the code".
Alternatively it might be possible to let actions reformat the code and push the changes to the PR.

**Comment by fischerscode** [comment]
> What does the code analysis need?

I think it needs all dependencies being resolved, since it also checks if the types match.
For the flutter package this means running pub get in the dart package is required.

**Comment by mtrezza** [comment]
I see; for npm lint it is not always able to auto-fix everything, so it's essential to see the exact error. But if this can auto-fix everything then I guess we are good here.

**Comment by fischerscode** [comment]
@mtrezza, pub get is also needed in the package where the tests are run. Just tested it.

**Comment by fischerscode** [comment]
> If we add a CONTRIBUTING.md file here with FAQ to explain what to do, then that'd also do it I think.

👍

**Comment by mtrezza** [comment]
Should I replace

```
(cd packages/dart && flutter pub get)
cd packages/flutter
```

with
```
(cd packages/flutter &&flutter pub get)
(cd packages/dart &&dart pub get)
```

**Comment by mtrezza** [comment]
> pub get is also needed in the package where the tests are run. Just tested it.

Could you do a review and let me know what I need to change in the review tab?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -11,52 +11,104 @@ on:
...
+        os: [ubuntu-latest, macos-latest, windows-latest]
+      fail-fast: false
+    name: Test dart (${{ matrix.os }})
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+      - name: Setup dart
+        uses: dart-lang/setup-dart@v1
+        with:
+          sdk: 'stable'
```
Comment (fischerscode) [inline .github/workflows/ci.yml:108]: stable should be the default, but specifying it explicitly should be fine, too.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -11,52 +11,104 @@ on:
...
+    name: Test flutter (${{ matrix.os }})
     steps:
-      - name: Checkout the code
+      - name: Checkout code
         uses: actions/checkout@v3
       - name: Setup flutter
         uses: subosito/flutter-action@v2
         with:
-          channel: ${{ matrix.channel }}
+          channel: 'stable'
```
Comment (fischerscode) [inline .github/workflows/ci.yml:84]: stable should be the default, but specifying it explicitly should be fine, too.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -11,52 +11,104 @@ on:
...
+    name: Code analysis (dart)
+    timeout-minutes: 5
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+      - name: Setup dart
+        uses: dart-lang/setup-dart@v1
+      - name: Analyze code
         run: dart analyze packages/dart
```
Comment (fischerscode) [inline .github/workflows/ci.yml:70]: `dart pub get --directory packages/dart` must be run before analyze

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -11,52 +11,104 @@ on:
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
         uses: actions/checkout@v3
       - name: Setup dart
         uses: dart-lang/setup-dart@v1
-        with:
-          sdk: ${{ matrix.sdk }}
       - name: Install dependencies
         run: dart pub get --directory packages/dart
```
Comment (fischerscode) [inline .github/workflows/ci.yml:37]: That should not be needed.

**Review by fischerscode — CHANGES_REQUESTED** [review]
Almost ready 👍

File: .github/workflows/ci.yml (hunk 4 above)
Comment (mtrezza) [reply .github/workflows/ci.yml]: removed

File: .github/workflows/ci.yml (hunk 5 above)
Comment (mtrezza) [reply .github/workflows/ci.yml]: removed

File: .github/workflows/ci.yml (hunk 6 above)
Comment (mtrezza) [reply .github/workflows/ci.yml:70]: added

File: .github/workflows/ci.yml (hunk 7 above)
Comment (mtrezza) [reply .github/workflows/ci.yml:37]: removed

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -11,52 +11,101 @@ on:
...
+      - name: Checkout code
         uses: actions/checkout@v3
       - name: Setup flutter
         uses: subosito/flutter-action@v2
         with:
-          channel: ${{ matrix.channel }}
+          cache: true
       - name: Install dependencies
         run: |
           (cd packages/dart && flutter pub get)
```
Comment (fischerscode) [inline .github/workflows/ci.yml:87]: Just realized, that's not working on windows.
Windows does not change back in the previous directory like linux/macos does :(

Replacing this line with
```
cd packages/dart
flutter pub get
cd ../..
```
should work.

**Review by fischerscode — CHANGES_REQUESTED** [review]
By looking through the ci logs, I've discovered a problem regarding special behavior on windows. Apart from that, it looks good to me.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -11,52 +11,105 @@ on:
...
+        uses: actions/checkout@v3
+      - name: Setup flutter
+        uses: subosito/flutter-action@v2
         with:
-          sdk: ${{ matrix.sdk }}
+          cache: true
+      - name: Install dependencies
+        run: |
+          cd packages/dart
+          flutter pub get
```
Comment (mtrezza) [inline .github/workflows/ci.yml:52]: @fischerscode shouldn't this be `dart pub get`?

File: .github/workflows/ci.yml (hunk 8 above)
Comment (mtrezza) [reply .github/workflows/ci.yml]: changed

File: .github/workflows/ci.yml (hunk 9 above)
Comment (fischerscode) [reply .github/workflows/ci.yml:52]: It should work.
I think dart would work, too. But I'm not sure if subosito/flutter-action also installs the dart command.

Comment (mtrezza) [reply .github/workflows/ci.yml:52]: OK so I leave it as it is?

Comment (fischerscode) [reply .github/workflows/ci.yml:52]: I would leave it like this.

**Review by fischerscode — APPROVED** [review]
Looks good to me.

**Pull request merged**